src/options.cpp

```diff
@@ -0,0 +1,408 @@
+
+Options::Options() {
+  oauth_server_info = update_server_info = &default_server_info;
+}
+
+Options::~Options() {
+  if (oauth_server_info != &default_server_info) {
+    delete oauth_server_info;
+  }
+  if (update_server_info != &default_server_info) {
+    delete update_server_info;
+  }
+}
+
+/**
+ * boost::lexical_cast doesn't support "true"/"false" strings to boolean
+ */
+namespace boost {
+  template<>
+  bool lexical_cast<bool, std::string>(const std::string& arg) {
+    std::istringstream ss(arg);
+    bool b = false;
+    ss >> std::boolalpha >> b;
+    return b;
+  }
+}
+
+template <typename T>
+static T check_attributes(const char **attrs, const char *attribute_name, const T &default_value) {
+  if (attrs == NULL || attribute_name == NULL)
+    return default_value;
+  while (*attrs) {
+    if (!std::strcmp(attrs[0], attribute_name))
+    {
+      try {
+        return boost::lexical_cast<T>(std::string(attrs[1]));
+      } catch (boost::bad_lexical_cast &) {}
+    }
+    attrs += 2;
+  }
+  return default_value;
+}
+
+apidb_server_info *check_attributes(const char **attrs, apidb_server_info* default_info) {
+  if (attrs == NULL)
+    return NULL;
+  apidb_server_info *result = default_info;
+  if (result == NULL)
+    result = new apidb_server_info();
+  while (*attrs) {
+    if (!std::strcmp(attrs[0], "charset") && std::strcmp(attrs[1], "") != 0)
+      result->charset = attrs[1];
+    else if (!std::strcmp(attrs[0], "host") && std::strcmp(attrs[1], "") != 0)
+      result->host = attrs[1];
+    else if (!std::strcmp(attrs[0], "port") && std::strcmp(attrs[1], "") != 0)
+      result->port = attrs[1];
+    else if (!std::strcmp(attrs[0], "dbname") && std::strcmp(attrs[1], "") != 0)
+      result->dbname = attrs[1];
+    else if (!std::strcmp(attrs[0], "username") && std::strcmp(attrs[1], "") != 0)
+      result->username = attrs[1];
+    else if (!std::strcmp(attrs[0], "password") && std::strcmp(attrs[1], "") != 0)
+      result->password = attrs[1];
+    attrs += 2;
+  }
+  return result;
+}
+
+void Options::parse_file(const std::string &options_path) {
+  if (!options_path.empty() && std::ifstream(options_path).good())
+    xmlpp::SaxParser::parse_file(options_path);
+  else
+    logger::message("Options file not found, using defaults.");
+}
+
+std::string Options::get_connect_db_str(BackendType type) const {
+  //  Build the connection string.
+  std::ostringstream ostr;
+  ostr << "dbname=" << get_backend_dbname(type);
+  if (!get_backend_host(type).empty())
+    ostr << " host=" << get_backend_host(type);
+  if (!get_backend_username(type).empty())
+    ostr << " user=" << get_backend_username(type);
+  if (!get_backend_password(type).empty())
+    ostr << " password=" << get_backend_password(type);
+  if (!get_backend_port(type).empty())
+    ostr << " port=" << get_backend_port(type);
+  return ostr.str();
+}
+
+void Options::on_start_element(const char *element, const char **attrs) {
```

Comment:
I left out all XML processing and let boost program options do the work instead. This not only saves us some extra parsing logic, but also makes sure that command line, environment and configuration file use the same parameter definition without a need to keep multiple places in sync.